src/tiledb/cloud/utilities/consolidate.py

```diff
@@ -0,0 +1,248 @@
+
+    logger = get_logger()
+
+    with tiledb.scope_ctx(config):
+        if vacuum_fragments:
+            try:
+                logger.info("Vacuuming fragments")
+                tiledb.vacuum(
+                    array_uri,
+                    config=tiledb.Config({"sm.vacuum.mode": "fragments"}),
+                )
+            except:
+                raise
+
+        # Modes for consolidate and vacuum
+        modes = ["commits", "fragment_meta"]
+
+        for mode in modes:
+            try:
+                logger.info(f"Consolidating {mode}")
+                tiledb.consolidate(
+                    array_uri,
+                    config=tiledb.Config({"sm.consolidation.mode": mode}),
+                )
+            except:
+                raise
+
+        for mode in modes:
+            try:
+                logger.info(f"Vacuuming {mode}")
+                tiledb.vacuum(
+                    array_uri,
+                    config=tiledb.Config({"sm.vacuum.mode": mode}),
+                )
+            except:
+                raise
+
+    logger.info(f"max memory usage: {max_memory_usage() / (1 << 30):.3f} GiB")
+
+
+def consolidate_fragments(
+    array_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    group_first_dim: bool = False,
+    graph: Optional[dag.DAG] = None,
+    dependencies: Optional[Sequence[dag.Node]] = None,
+    consolidate_resources: Optional[Mapping[str, str]] = None,
+    namespace: Optional[str] = None,
+    max_fragment_size: int = MAX_FRAGMENT_SIZE,
+) -> None:
+    """
+    Consolidate fragments in an array.
+
+    If `group_first_dim` is True, fragments with the same value for the first dimension
+    will be consolidated together. Otherwise, all fragments will be consolidated
+    together.
+
+    If `graph` is provided, the consolidation task nodes will be submitted to the graph.
+    If `dependencies` is provided, the consolidation nodes will depend on the nodes in
+    the list.
+
+    If `graph` is not provided, a new graph will be created and submitted to TileDB
+    Cloud.
+
+    :param array_uri: array URI
+    :param config: config dictionary, defaults to None
+    :param group_first_dim: group fragment by first dimension, defaults to True
+    :param graph: graph to submit nodes to, defaults to None
+    :param dependencies: list of nodes in the graph to depend on, defaults to None
+    :param consolidate_resources: resources for the consolidate node, defaults to None
+    :param namespace: TileDB Cloud namespace, defaults to the user's default namespace
+    :param max_fragment_size: max size of consolidated fragments,
+        defaults to MAX_FRAGMENT_SIZE
+    """
+
+    if graph is None and dependencies is not None:
+        raise ValueError("Graph must be provided if dependencies are provided")
+
+    graph_provided = graph is not None
+
+    # If a graph is not provided, create a new graph and run it at the end of this
+    # function.
+    if not graph_provided:
```

Comment:
maybe make it `graph_omitted` so that you can use it in `if graph_omitted and dependencies is not None` to avoid some minor repetition.